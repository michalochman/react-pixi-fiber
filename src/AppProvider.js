import React from "react";
import PropTypes from "prop-types";
import { isNewContextAvailable } from "./compat";

const propTypes = {
  app: PropTypes.object.isRequired,
  children: PropTypes.node,
};
const childContextTypes = {
  app: PropTypes.object,
};

let AppContext = null;

function createAppProvider() {
  if (isNewContextAvailable()) {
    // New Context API
    if (AppContext === null) {
      AppContext = React.createContext(null);
    }

    class AppProvider extends React.Component {
      render() {
        const { app, children } = this.props;
        return <AppContext.Provider value={app}>{children}</AppContext.Provider>;
      }
    }

    AppProvider.propTypes = propTypes;

    const withApp = WrappedComponent => {
      function WithApp(props) {
        return <AppContext.Consumer>{app => <WrappedComponent {...props} app={app} />}</AppContext.Consumer>;
      }
      WithApp.displayName = `withApp(${WrappedComponent})`;

      return WithApp;
    };

    return { AppProvider, withApp };
  } else {
    // Legacy Context API
    class AppProvider extends React.Component {
      getChildContext() {
        return {
          app: this.props.app,
        };
      }

      render() {
        return this.props.children;
      }
    }

    AppProvider.propTypes = propTypes;
    AppProvider.childContextTypes = childContextTypes;

    const withApp = WrappedComponent => {
      function WithApp(props, context) {
        return <WrappedComponent {...props} app={context.app} />;
      }
      WithApp.displayName = `withApp(${WrappedComponent})`;
      WithApp.contextTypes = childContextTypes;

      return WithApp;
    };

    return { AppProvider, withApp };
  }
}

const { AppProvider, withApp } = createAppProvider();

export { AppContext, AppProvider, withApp };
