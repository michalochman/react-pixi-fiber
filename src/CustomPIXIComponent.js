import invariant from "fbjs/lib/invariant";
import { injectType } from "./inject";

function CustomPIXIComponent(behavior, type) {
  invariant(
    typeof type === "string",
    "Invalid argument `type` of type `%s` supplied to `CustomPIXIComponent`, expected `string`.",
    typeof type
  );

  return injectType(type, behavior);
}

export default CustomPIXIComponent;
