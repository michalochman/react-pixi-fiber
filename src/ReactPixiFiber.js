import ReactFiberReconciler from "react-reconciler";
import emptyObject from "fbjs/lib/emptyObject";
import invariant from "fbjs/lib/invariant";
import now from "performance-now";
import * as PIXI from "pixi.js";
import { createInjectedTypeInstance, isInjectedType } from "./inject";
import { CHILDREN, DEFAULT_PROPS } from "./props";
import { TYPES } from "./types";
import { filterByKey, including, includingReservedProps, not, setPixiValue } from "./utils";

/* Render Methods */

// TODO consider whitelisting props based on component type
export function defaultApplyProps(instance, oldProps, newProps) {
  Object.keys(newProps)
    .filter(not(includingReservedProps))
    .forEach(propName => {
      const value = newProps[propName];

      // Set value if defined
      if (typeof value !== "undefined") {
        setPixiValue(instance, propName, value);
      } else if (typeof instance[propName] !== "undefined" && typeof DEFAULT_PROPS[propName] !== "undefined") {
        // Reset to default value (if it is defined) when display object had prop set and no longer has
        console.warn(`setting default value: ${propName} was ${instance[propName]} is ${value} for`, instance);
        setPixiValue(instance, propName, DEFAULT_PROPS[propName]);
      } else {
        console.warn(`ignoring prop: ${propName} was ${instance[propName]} is ${value} for`, instance);
      }
    });
}

export function applyProps(instance, oldProps, newProps) {
  if (typeof instance._customApplyProps === "function") {
    instance._customApplyProps(instance, oldProps, newProps);
  } else {
    defaultApplyProps(instance, oldProps, newProps);
  }
}

// Calculate the diff between the two objects.
// See: https://github.com/facebook/react/blob/97e2911/packages/react-dom/src/client/ReactDOMFiberComponent.js#L546
export function diffProps(pixiElement, type, lastRawProps, nextRawProps, rootContainerElement) {
  let updatePayload = null;

  let lastProps = lastRawProps;
  let nextProps = nextRawProps;
  let propKey;

  for (propKey in lastProps) {
    if (nextProps.hasOwnProperty(propKey) || !lastProps.hasOwnProperty(propKey) || lastProps[propKey] == null) {
      continue;
    }
    if (propKey === CHILDREN) {
      // Noop. Text children not supported
    } else {
      // For all other deleted properties we add it to the queue. We use
      // the whitelist in the commit phase instead.
      (updatePayload = updatePayload || []).push(propKey, null);
    }
  }
  for (propKey in nextProps) {
    const nextProp = nextProps[propKey];
    const lastProp = lastProps != null ? lastProps[propKey] : undefined;
    if (!nextProps.hasOwnProperty(propKey) || nextProp === lastProp || (nextProp == null && lastProp == null)) {
      continue;
    }
    if (propKey === CHILDREN) {
      // Noop. Text children not supported
    } else {
      // For any other property we always add it to the queue and then we
      // filter it out using the whitelist during the commit.
      (updatePayload = updatePayload || []).push(propKey, nextProp);
    }
  }
  return updatePayload;
}

/* PixiJS Renderer */

export function appendChild(parentInstance, child) {
  // TODO do we need to remove the child first if it's already added?
  parentInstance.removeChild(child);

  parentInstance.addChild(child);
  if (typeof child._customDidAttach === "function") {
    child._customDidAttach(child);
  }
}

export function removeChild(parentInstance, child) {
  if (typeof child._customWillDetach === "function") {
    child._customWillDetach(child);
  }

  parentInstance.removeChild(child);

  child.destroy();
}

export function insertBefore(parentInstance, child, beforeChild) {
  invariant(child !== beforeChild, "ReactPixiFiber cannot insert node before itself");

  const childExists = parentInstance.children.indexOf(child) !== -1;
  const index = parentInstance.getChildIndex(beforeChild);

  if (childExists) {
    parentInstance.setChildIndex(child, index);
  } else {
    parentInstance.addChildAt(child, index);
  }
}

export function commitUpdate(instance, updatePayload, type, lastRawProps, nextRawProps, internalInstanceHandle) {
  // injected types need to have full control over passed props
  if (isInjectedType(type)) {
    applyProps(instance, lastRawProps, nextRawProps);
    return;
  }

  // updatePayload is in the form of [propKey1, propValue1, ...]
  const updatedPropKeys = including(updatePayload.filter((item, i) => i % 2 === 0));
  const oldProps = filterByKey(lastRawProps, updatedPropKeys);
  const newProps = filterByKey(nextRawProps, updatedPropKeys);

  // regular components only receive props that have changed
  applyProps(instance, oldProps, newProps);
}

export function createInstance(type, props, internalInstanceHandle) {
  let instance;

  switch (type) {
    case TYPES.BITMAP_TEXT:
      instance = new PIXI.extras.BitmapText(props.text, props.style);
      break;
    case TYPES.CONTAINER:
      instance = new PIXI.Container();
      break;
    case TYPES.GRAPHICS:
      instance = new PIXI.Graphics();
      break;
    case TYPES.PARTICLE_CONTAINER:
      instance = new PIXI.particles.ParticleContainer(
        props.maxSize,
        props.properties,
        props.batchSize,
        props.autoResize
      );
      break;
    case TYPES.SPRITE:
      instance = new PIXI.Sprite(props.texture);
      break;
    case TYPES.TEXT:
      instance = new PIXI.Text(props.text, props.style, props.canvas);
      break;
    case TYPES.TILING_SPRITE:
      instance = new PIXI.extras.TilingSprite(props.texture, props.width, props.height);
      break;
    default:
      instance = createInjectedTypeInstance(type, props, internalInstanceHandle, defaultApplyProps);
      break;
  }

  invariant(instance, "ReactPixiFiber does not support the type: `%s`.", type);

  applyProps(instance, {}, props);

  return instance;
}

export function createTextInstance(text, rootContainerInstance, internalInstanceHandle) {
  invariant(false, "ReactPixiFiber does not support text instances. Use Text component instead.");
}

export function finalizeInitialChildren(pixiElement, type, props, rootContainerInstance, hostContext) {
  return false;
}

export function getChildHostContext(parentHostContext, type) {
  return emptyObject;
}

export function getRootHostContext(rootContainerInstance) {
  return emptyObject;
}

export function getPublicInstance(inst) {
  return inst;
}

export function prepareForCommit() {
  // Noop
}

export function prepareUpdate(pixiElement, type, oldProps, newProps, rootContainerInstance, hostContext) {
  return diffProps(pixiElement, type, oldProps, newProps, rootContainerInstance);
}

export function resetAfterCommit() {
  // Noop
}

export function resetTextContent(pixiElement) {
  // Noop
}

export function shouldDeprioritizeSubtree(type, props) {
  const isAlphaVisible = typeof props.alpha === "undefined" || props.alpha > 0;
  const isRenderable = typeof props.renderable === "undefined" || props.renderable === true;
  const isVisible = typeof props.visible === "undefined" || props.visible === true;

  return !(isAlphaVisible && isRenderable && isVisible);
}

export function shouldSetTextContent(type, props) {
  return false;
}

export function commitTextUpdate(textInstance, oldText, newText) {
  // Noop
}

export function commitMount(instance, type, newProps) {
  // Noop
}

const ReactPixiFiber = ReactFiberReconciler({
  appendInitialChild: appendChild,
  createInstance: createInstance,
  createTextInstance: createTextInstance,
  finalizeInitialChildren: finalizeInitialChildren,
  getChildHostContext: getChildHostContext,
  getRootHostContext: getRootHostContext,
  getPublicInstance: getPublicInstance,
  now: now,
  prepareForCommit: prepareForCommit,
  prepareUpdate: prepareUpdate,
  resetAfterCommit: resetAfterCommit,
  resetTextContent: resetTextContent,
  shouldDeprioritizeSubtree: shouldDeprioritizeSubtree,
  shouldSetTextContent: shouldSetTextContent,
  mutation: {
    appendChild: appendChild,
    appendChildToContainer: appendChild,
    commitMount: commitMount,
    commitTextUpdate: commitTextUpdate,
    commitUpdate: commitUpdate,
    insertBefore: insertBefore,
    insertInContainerBefore: insertBefore,
    removeChild: removeChild,
    removeChildFromContainer: removeChild,
  },
});

export default ReactPixiFiber;
