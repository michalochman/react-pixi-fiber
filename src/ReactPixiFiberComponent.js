// Based on: https://github.com/facebook/react/blob/27535e7bfcb63e8a4d65f273311e380b4ca12eff/packages/react-dom/src/client/ReactDOMFiberComponent.js
import invariant from "fbjs/lib/invariant";
import * as PIXI from "pixi.js";
import { CHILDREN } from "./props";
import { TYPES } from "./types";
import { createInjectedTypeInstance, isInjectedType } from "./inject";
import { setValueForProperty } from "./PixiPropertyOperations";
import { validateProperties as validateUnknownProperties } from "./ReactPixiFiberUnknownPropertyHook";

let validatePropertiesInDevelopment;

if (__DEV__) {
  validatePropertiesInDevelopment = function(type, props) {
    validateUnknownProperties(type, props);
  };
}

export function createInstance(type, props, internalInstanceHandle) {
  let instance;

  switch (type) {
    case TYPES.BITMAP_TEXT:
      try {
        instance = new PIXI.extras.BitmapText(props.text, props.style);
      } catch (e) {
        instance = new PIXI.BitmapText(props.text, props.style);
      }
      break;
    case TYPES.CONTAINER:
      instance = new PIXI.Container();
      break;
    case TYPES.GRAPHICS:
      instance = new PIXI.Graphics();
      break;
    case TYPES.PARTICLE_CONTAINER:
      try {
        instance = new PIXI.particles.ParticleContainer(
          props.maxSize,
          props.properties,
          props.batchSize,
          props.autoResize
        );
      } catch (e) {
        instance = new PIXI.ParticleContainer(props.maxSize, props.properties, props.batchSize, props.autoResize);
      }
      break;
    case TYPES.SPRITE:
      instance = new PIXI.Sprite(props.texture);
      break;
    case TYPES.TEXT:
      instance = new PIXI.Text(props.text, props.style, props.canvas);
      break;
    case TYPES.TILING_SPRITE:
      try {
        instance = new PIXI.extras.TilingSprite(props.texture, props.width, props.height);
      } catch (e) {
        instance = new PIXI.TilingSprite(props.texture, props.width, props.height);
      }
      break;
    default:
      instance = createInjectedTypeInstance(type, props, internalInstanceHandle, applyDisplayObjectProps);
      break;
  }

  invariant(instance, "ReactPixiFiber does not support the type: `%s`.", type);

  return instance;
}

export function setInitialCustomComponentProperties(type, instance, rawProps, rootContainerElement) {
  instance._customApplyProps(instance, undefined, rawProps);
}

export function setInitialPixiProperties(type, instance, rawProps, rootContainerElement) {
  for (const propKey in rawProps) {
    if (!rawProps.hasOwnProperty(propKey)) {
      continue;
    }
    const nextProp = rawProps[propKey];
    if (propKey === CHILDREN) {
      // Noop. Text children not supported
    } else {
      setValueForProperty(type, instance, propKey, nextProp);
    }
  }
}

export function setInitialProperties(type, instance, rawProps, rootContainerElement) {
  // injected types with customApplyProps need to have full control over passed props
  if (isInjectedType(type) && typeof instance._customApplyProps === "function") {
    setInitialCustomComponentProperties(type, instance, rawProps, rootContainerElement);
    return;
  }

  if (__DEV__) {
    validatePropertiesInDevelopment(type, rawProps);
  }

  setInitialPixiProperties(type, instance, rawProps, rootContainerElement);
}

// Calculate the diff between the two objects.
// See: https://github.com/facebook/react/blob/97e2911/packages/react-dom/src/client/ReactDOMFiberComponent.js#L546
export function diffProperties(type, instance, lastRawProps, nextRawProps, rootContainerElement) {
  if (__DEV__) {
    validatePropertiesInDevelopment(type, nextRawProps);
  }

  let updatePayload = null;

  let lastProps = lastRawProps;
  let nextProps = nextRawProps;
  let propKey;

  for (propKey in lastProps) {
    if (nextProps.hasOwnProperty(propKey) || !lastProps.hasOwnProperty(propKey)) {
      continue;
    }
    if (propKey === CHILDREN) {
      // Noop. Text children not supported
    } else {
      // For all other deleted properties we add it to the queue. We use
      // the whitelist in the commit phase instead.
      (updatePayload = updatePayload || []).push(propKey, null);
    }
  }
  for (propKey in nextProps) {
    const nextProp = nextProps[propKey];
    const lastProp = lastProps != null ? lastProps[propKey] : undefined;
    if (!nextProps.hasOwnProperty(propKey) || nextProp === lastProp) {
      continue;
    }
    if (propKey === CHILDREN) {
      // Noop. Text children not supported
    } else {
      // For any other property we always add it to the queue and then we
      // filter it out using the whitelist during the commit.
      (updatePayload = updatePayload || []).push(propKey, nextProp);
    }
  }
  return updatePayload;
}

export function applyDisplayObjectProps(type, instance, oldProps, newProps) {
  const updatePayload = diffProperties(type, instance, oldProps, newProps);
  if (updatePayload !== null) {
    updatePixiProperties(type, instance, updatePayload);
  }
}

export function updateCustomComponentProperties(
  type,
  instance,
  updatePayload,
  lastRawProps,
  nextRawProps,
  internalInstanceHandle
) {
  instance._customApplyProps(instance, lastRawProps, nextRawProps);
}

export function updatePixiProperties(
  type,
  instance,
  updatePayload,
  lastRawProps,
  nextRawProps,
  internalInstanceHandle
) {
  for (let i = 0; i < updatePayload.length; i += 2) {
    const propKey = updatePayload[i];
    const propValue = updatePayload[i + 1];
    if (propKey === CHILDREN) {
      // Noop. Text children not supported
    } else {
      setValueForProperty(type, instance, propKey, propValue);
    }
  }
}

// Apply the diff.
export function updateProperties(type, instance, updatePayload, lastRawProps, nextRawProps, internalInstanceHandle) {
  // injected types with customApplyProps need to have full control over passed props
  if (isInjectedType(type) && typeof instance._customApplyProps === "function") {
    updateCustomComponentProperties(type, instance, updatePayload, lastRawProps, nextRawProps, internalInstanceHandle);
    return;
  }

  updatePixiProperties(type, instance, updatePayload, lastRawProps, nextRawProps, internalInstanceHandle);
}
