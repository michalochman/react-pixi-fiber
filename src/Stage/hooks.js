import React, { useEffect, useLayoutEffect, useRef, useState } from "react";
import { cleanup, renderStage, rerenderStage, resizeRenderer } from "./common";
import { defaultProps, getCanvasProps, propTypes } from "./propTypes";
import { createPixiApplication } from "../utils";

export function usePixiAppCreator(props) {
  const { options } = props;
  const { height, width, ...otherOptions } = options;

  const canvasRef = useRef();
  const [app, setApp] = useState(null);

  const canvasProps = getCanvasProps(props);
  // Do not render anything if view is passed to options
  // TODO should it return options.view instead of null?
  const canvas = options && options.view ? null : <canvas ref={canvasRef} {...canvasProps} />;

  // Initialize pixi application on mount
  useLayoutEffect(
    () => {
      const view = canvasRef.current;
      const appInstance = createPixiApplication({ view, ...options });

      setApp(appInstance);

      // Destroy pixi application on unmount
      return () => {
        const removeView = options.view;

        // Destroy PIXI.Application and remove canvas if necessary
        cleanup(appInstance, removeView);
      };
    },
    // We need to create new Application when options other than dimensions
    // are changed because some of the renderer settings are immutable
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [...Object.values(otherOptions)]
  );

  return { app, canvas };
}

export function usePreviousProps(value) {
  const ref = useRef({});

  useEffect(() => {
    ref.current = value;
  });

  return ref.current;
}

// Render stage for the first time
export function useStageRenderer(app, props) {
  useLayoutEffect(
    () => {
      if (!app || !app.stage) return;

      renderStage(app, props);
    },
    // Only act if the app was created
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [app]
  );
}

// Re-render stage when component updates
export function useStageRerenderer(app, props) {
  const prevProps = usePreviousProps(props);

  useLayoutEffect(
    () => {
      if (!app || !app.stage) return;

      rerenderStage(app, prevProps, props);
    },
    // Only act if new app was created or props have changed
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [app, props]
  );
}

// Root container might have been resized - resize renderer
export function useRendererResizer(app, props) {
  const prevProps = usePreviousProps(props);

  useLayoutEffect(
    () => {
      if (!app || !app.renderer) return;

      resizeRenderer(app, props, prevProps);
    },
    // Only act if new app was created or props have changed
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [app, props]
  );
}

export default function createStageFunction() {
  function Stage(props) {
    const { app, canvas } = usePixiAppCreator(props);

    useStageRenderer(app, props);
    useStageRerenderer(app, props);
    useRendererResizer(app, props);

    return canvas;
  }

  Stage.propTypes = propTypes;
  Stage.defaultProps = defaultProps;

  return Stage;
}
