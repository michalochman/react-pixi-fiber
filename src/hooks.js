import React from "react";
import { useContext, useEffect } from "react";
import { AppContext } from "./AppProvider";

export function usePixiApp() {
  return useContext(AppContext);
}

export function usePixiTicker(fn) {
  const { ticker } = usePixiApp();

  useEffect(() => {
    ticker.add(fn);

    return () => {
      ticker.remove(fn);
    };
  }, [fn, ticker]);
}
