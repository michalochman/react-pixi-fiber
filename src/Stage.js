import React from "react";
import PropTypes from "prop-types";
import * as PIXI from "pixi.js";
import pkg from "../package.json";
import { DEFAULT_PROPS } from "./props";
import ReactPixiFiber, { applyProps } from "./ReactPixiFiber";
import { filterByKey, including } from "./utils";

export function validateCanvas(props, propName, componentName) {
  // Let's assume that element is canvas if the element is Element and implements getContext
  const element = props[propName];
  if (typeof element === "undefined") {
    return;
  }

  const isCanvas = element instanceof Element && typeof element.getContext === "function";
  if (!isCanvas) {
    const propType = typeof element;
    return new Error(
      `Invalid prop '${propName}' of type '${propType}' supplied to '${componentName}', expected '<canvas> Element'.`
    );
  }
}

const propTypes = {
  options: PropTypes.shape({
    antialias: PropTypes.bool,
    autoStart: PropTypes.bool,
    backgroundColor: PropTypes.number,
    clearBeforeRender: PropTypes.bool,
    forceCanvas: PropTypes.bool,
    forceFXAA: PropTypes.bool,
    height: PropTypes.number,
    legacy: PropTypes.bool,
    powerPreference: PropTypes.string,
    preserveDrawingBuffer: PropTypes.bool,
    resolution: PropTypes.number,
    roundPixels: PropTypes.bool,
    sharedLoader: PropTypes.bool,
    sharedTicker: PropTypes.bool,
    transparent: PropTypes.bool,
    view: validateCanvas,
    width: PropTypes.number,
  }),
  children: PropTypes.node,
  height: PropTypes.number,
  width: PropTypes.number,
};
const childContextTypes = {
  app: PropTypes.object,
};

export const includingDisplayObjectProps = including(Object.keys(DEFAULT_PROPS));
export const includingStageProps = including(Object.keys(propTypes));
export const includingCanvasProps = key => !includingDisplayObjectProps(key) && !includingStageProps(key);

export const getCanvasProps = props => filterByKey(props, includingCanvasProps);
export const getDisplayObjectProps = props => filterByKey(props, includingDisplayObjectProps);

class Stage extends React.Component {
  getChildContext() {
    return {
      app: this._app,
    };
  }

  componentDidMount() {
    const { children, height, options, width } = this.props;

    this._app = new PIXI.Application(width, height, {
      view: this._canvas,
      ...options,
    });

    // Apply root Container props
    const stageProps = getDisplayObjectProps(this.props);
    applyProps(this._app.stage, {}, stageProps);

    // Perhaps this should use the standalone render method somehow, the only differences now are:
    // - parentContainer
    // - callback
    // - return value
    this._mountNode = ReactPixiFiber.createContainer(this._app.stage);
    ReactPixiFiber.updateContainer(children, this._mountNode, this);

    ReactPixiFiber.injectIntoDevTools({
      findFiberByHostInstance: ReactPixiFiber.findFiberByHostInstance,
      bundleType: __DEV__ ? 1 : 0,
      version: pkg.version,
      rendererPackageName: pkg.name,
    });
  }

  componentDidUpdate(prevProps, prevState) {
    const { children, height, width } = this.props;

    // Apply root Container props
    const stageProps = getDisplayObjectProps(this.props);
    applyProps(this._app.stage, {}, stageProps);

    // Root container has been resized - resize renderer
    if (height !== prevProps.height || width !== prevProps.width) {
      this._app.renderer.resize(width, height);
    }

    ReactPixiFiber.updateContainer(children, this._mountNode, this);
  }

  componentWillUnmount() {
    ReactPixiFiber.updateContainer(null, this._mountNode, this);
  }

  render() {
    const { options } = this.props;
    const canvasProps = getCanvasProps(this.props);

    // Do not render anything if view is passed to options
    if (typeof options !== "undefined" && options.view) {
      return null;
    } else {
      return <canvas ref={ref => (this._canvas = ref)} {...canvasProps} />;
    }
  }
}

Stage.propTypes = propTypes;
Stage.childContextTypes = childContextTypes;

export default Stage;
