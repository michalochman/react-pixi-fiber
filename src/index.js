import CustomPIXIComponent from "./CustomPIXIComponent";
import { AppContext, AppProvider, withApp } from "./AppProvider";
import Stage, { createStageClass } from "./Stage";
import { TYPES } from "./types";
import { usePixiApp, usePixiTicker } from "./hooks";
import { createRender, createUnmount } from "./render";
import { ReactPixiFiberAsPrimaryRenderer, unstable_batchedUpdates } from "./ReactPixiFiber";
import { applyDisplayObjectProps } from "./ReactPixiFiberComponent";

const render = createRender(ReactPixiFiberAsPrimaryRenderer);
const unmount = createUnmount(ReactPixiFiberAsPrimaryRenderer);

/* Public API */

export {
  AppContext,
  AppProvider,
  CustomPIXIComponent,
  Stage,
  applyDisplayObjectProps,
  createStageClass,
  render,
  unmount,
  withApp,
  usePixiApp,
  usePixiTicker,
  unstable_batchedUpdates,
};

export const BitmapText = TYPES.BITMAP_TEXT;
export const Container = TYPES.CONTAINER;
export const Graphics = TYPES.GRAPHICS;
export const ParticleContainer = TYPES.PARTICLE_CONTAINER;
export const Sprite = TYPES.SPRITE;
export const Text = TYPES.TEXT;
export const TilingSprite = TYPES.TILING_SPRITE;
