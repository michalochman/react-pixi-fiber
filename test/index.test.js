import * as ReactPixiFiber from "../src/index";
import CustomPIXIComponent from "../src/CustomPIXIComponent";
import { AppContext, AppProvider, withApp } from "../src/AppProvider";
import Stage, { createStageClass } from "../src/Stage";
import { TYPES } from "../src/types";
import { usePixiApp, usePixiTicker } from "../src/hooks";
import { unstable_batchedUpdates } from "../src/ReactPixiFiber";
import { applyDisplayObjectProps } from "../src/ReactPixiFiberComponent";

describe("ReactPixiFiber public API", () => {
  it("should match snapshot", () => {
    expect(ReactPixiFiber).toMatchSnapshot();
  });

  it("provides expected utils", () => {
    expect(ReactPixiFiber.CustomPIXIComponent).toEqual(CustomPIXIComponent);
    expect(ReactPixiFiber.applyDisplayObjectProps).toEqual(applyDisplayObjectProps);
    expect(ReactPixiFiber.createStageClass).toEqual(createStageClass);
    expect(typeof ReactPixiFiber.render).toEqual("function");
    expect(typeof ReactPixiFiber.unmount).toEqual("function");
    expect(ReactPixiFiber.unstable_batchedUpdates).toEqual(unstable_batchedUpdates);
  });

  it("provides expected context utils", () => {
    expect(ReactPixiFiber.AppContext).toEqual(AppContext);
    expect(ReactPixiFiber.AppProvider).toEqual(AppProvider);
    expect(ReactPixiFiber.withApp).toEqual(withApp);
  });

  it("provides expected hooks", () => {
    expect(ReactPixiFiber.usePixiApp).toEqual(usePixiApp);
    expect(ReactPixiFiber.usePixiTicker).toEqual(usePixiTicker);
  });

  it("provides expected components", () => {
    expect(ReactPixiFiber.BitmapText).toEqual(TYPES.BITMAP_TEXT);
    expect(ReactPixiFiber.Container).toEqual(TYPES.CONTAINER);
    expect(ReactPixiFiber.Graphics).toEqual(TYPES.GRAPHICS);
    expect(ReactPixiFiber.ParticleContainer).toEqual(TYPES.PARTICLE_CONTAINER);
    expect(ReactPixiFiber.Sprite).toEqual(TYPES.SPRITE);
    expect(ReactPixiFiber.Stage).toEqual(Stage);
    expect(ReactPixiFiber.Text).toEqual(TYPES.TEXT);
    expect(ReactPixiFiber.TilingSprite).toEqual(TYPES.TILING_SPRITE);
  });
});
