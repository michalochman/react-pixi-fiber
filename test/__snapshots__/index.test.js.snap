// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactPixiFiber public API should match snapshot 1`] = `
Object {
  "AppContext": Object {
    "$$typeof": Symbol(react.context),
    "Consumer": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": [Circular],
    },
    "Provider": Object {
      "$$typeof": Symbol(react.provider),
      "_context": [Circular],
    },
    "_calculateChangedBits": null,
    "_currentRenderer": null,
    "_currentRenderer2": null,
    "_currentValue": null,
    "_currentValue2": null,
    "_threadCount": 0,
  },
  "AppProvider": [Function],
  "BitmapText": "BitmapText",
  "Container": "Container",
  "CustomPIXIComponent": [Function],
  "Graphics": "Graphics",
  "ParticleContainer": "ParticleContainer",
  "Sprite": "Sprite",
  "Stage": [Function],
  "Text": "Text",
  "TilingSprite": "TilingSprite",
  "applyDisplayObjectProps": [Function],
  "createStageClass": [Function],
  "default": Object {},
  "render": [Function],
  "unmount": [Function],
  "unstable_batchedUpdates": [Function],
  "usePixiApp": [Function],
  "usePixiTicker": [Function],
  "withApp": [Function],
}
`;
